prow/jira/jira.go

```diff
@@ -0,0 +1,232 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package jira
+
+import (
+	"errors"
+	"fmt"
+	"net/http"
+	"net/url"
+	"strings"
+
+	"github.com/andygrunwald/go-jira"
+	"github.com/hashicorp/go-retryablehttp"
+	"github.com/sirupsen/logrus"
+	"k8s.io/apimachinery/pkg/util/sets"
+)
+
+type Client interface {
+	GetIssue(id string) (*jira.Issue, error)
+	GetRemoteLinks(id string) ([]jira.RemoteLink, error)
+	AddRemoteLink(id string, link *jira.RemoteLink) error
+	JiraURL() string
+}
+
+type BasicAuthGenerator func() (username, password string)
+
+type Options struct {
+	BasicAuth BasicAuthGenerator
+	LogFields logrus.Fields
+}
+
+type Option func(*Options)
+
+func WithBasicAuth(basicAuth BasicAuthGenerator) Option {
+	return func(o *Options) {
+		o.BasicAuth = basicAuth
+	}
+}
+
+func WithFields(fields logrus.Fields) Option {
+	return func(o *Options) {
+		o.LogFields = fields
+	}
+}
+
+func NewClient(endpoint string, opts ...Option) (Client, error) {
```

Comment:
Can you copy the logger/delegate design from the GitHub, Bugzilla clients so that we can inject event-specific loggers into these clients when we're using them in a `hook` goroutine?

Reply:
This is pretty tricky, because the only place where we actually use the logger is in the `retryable` http client. This in turn is wrapped with the basic auth transport and we can not replace the transport on an existing `*jiraClient`. So the only way I can think of that would allow us to make this happen is by going through the constructor again. Unfortunately, the constructor can error which makes this hard to use.

Reply:
Why not log the HTTP calls the way we do for other clients?